src/storage/kv/transaction.rs

```diff
@@ -0,0 +1,1163 @@
+            let val2 = txn3.get(&key2).unwrap();
+            assert_eq!(val2, value2.as_ref());
+        }
+    }
+
+    // G1b: Intermediate Reads (dirty reads)
+    #[test]
+    fn test_snapshot_isolation_g1b() {
+        let store = create_store();
+
+        let key1 = Bytes::from("k1");
+        let key2 = Bytes::from("k2");
+        let value1 = Bytes::from("v1");
+        let value3 = Bytes::from("v3");
+        let value4 = Bytes::from("v4");
+
+        {
+            let mut txn1 = store.begin().unwrap();
+            let mut txn2 = store.begin().unwrap();
+
+            assert!(txn1.get(&key1).is_ok());
+            assert!(txn1.get(&key2).is_ok());
+
+            txn1.set(&key1, &value3).unwrap();
+
+            let range = VectorKey::from_str("k1")..=VectorKey::from_str("k3");
+            let res = txn2.scan(range.clone()).unwrap();
+            assert_eq!(res.len(), 2);
+            assert_eq!(res[0].0, value1);
+
+            txn1.set(&key1, &value4).unwrap();
+            txn1.commit().unwrap();
+
+            let res = txn2.scan(range).unwrap();
+            assert_eq!(res.len(), 2);
+            assert_eq!(res[0].0, value1);
+
+            txn2.commit().unwrap();
+        }
+    }
+
+    // G1c: Circular Information Flow (dirty reads)
+    #[test]
+    fn test_snapshot_isolation_g1c() {
+        let store = create_store();
+
+        let key1 = Bytes::from("k1");
+        let key2 = Bytes::from("k2");
+        let value1 = Bytes::from("v1");
+        let value2 = Bytes::from("v2");
+        let value3 = Bytes::from("v3");
+        let value4 = Bytes::from("v4");
+
+        {
+            let mut txn1 = store.begin().unwrap();
+            let mut txn2 = store.begin().unwrap();
+
+            assert!(txn1.get(&key1).is_ok());
+            assert!(txn2.get(&key2).is_ok());
+
+            txn1.set(&key1, &value3).unwrap();
+            txn2.set(&key2, &value4).unwrap();
+
+            assert_eq!(txn1.get(&key2).unwrap(), value2.as_ref());
+            assert_eq!(txn2.get(&key1).unwrap(), value1.as_ref());
+
+            txn1.commit().unwrap();
+            assert!(match txn2.commit() {
+                Err(err) => {
+                    if let Error::TxnReadConflict = err {
+                        true
+                    } else {
+                        false
+                    }
+                }
+                _ => false,
+            });
+        }
+    }
+
+    // PMP: Predicate-Many-Preceders
+    #[test]
+    fn test_snapshot_isolation_pmp() {
+        let store = create_store();
+
+        let key2 = Bytes::from("k2");
+        let key3 = Bytes::from("k3");
+        let value3 = Bytes::from("v3");
+
+        {
+            let txn1 = store.begin().unwrap();
+            let mut txn2 = store.begin().unwrap();
+
+            assert!(txn1.get(&key3).is_err());
+            txn2.set(&key2, &value3).unwrap();
+            txn2.commit().unwrap();
+
+            assert!(txn1.get(&key3).is_err());
```

Comment:
(Sorry if I misunderstood something badly, but) I think we want to avoid this anomaly either (1)by failing a read when it is about to return the data that causes a pmp anomaly OR (2)by returning the data from the snapshot associated to the tx.

Which behavior are you trying to check with this test?

Reply:
Thanks for the comment @mumoshu 

In a transactional context, pmp here is refering to a situation to test if multiple transactions are well isolated from each other, and read predicates are isolated.

Given the two transactions, what I am trying to check is that even though txn2 is committed, txn1 still has its own snapshot and won't read data which has been committed post it began.

Does that answer your question? Please feel free to lemme know. Ideally this fails in read committed or other isolation levels. Example [here](https://github.com/ept/hermitage/blob/master/sqlserver.md)

Reply:
> what I am trying to check is that even though txn2 is committed, txn1 still has its own snapshot and won't read data which has been committed post it began.

Thanks! But txn2 writes to key2 only and tx1 reads from key3 == no intersection of reads and writes. So, this test won't fail even if it had such anomaly in reality, no?

More specifically, this test would fail only when the datastore confuses writes to key2 as writes to key3, which is something worse than the anomaly we'd like to detect here...?